src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    /// <summary>
+    /// Represents a read-only collection of <typeparamref name="TAttribute"/>s.
+    /// </summary>
+    /// <typeparam name="TAttribute">
+    /// The type of <see cref="IReadOnlyTagHelperAttribute"/>s in the collection.
+    /// </typeparam>
+    public class ReadOnlyTagHelperAttributes<TAttribute> : IReadOnlyList<TAttribute>
+        where TAttribute : IReadOnlyTagHelperAttribute
+    {
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with an empty
+        /// collection.
+        /// </summary>
+        protected ReadOnlyTagHelperAttributes()
+        {
+            Attributes = new List<TAttribute>();
+        }
+
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with the specified
+        /// <paramref name="attributes"/>.
+        /// </summary>
+        /// <param name="attributes">The collection to wrap.</param>
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
+        {
+            Attributes = new List<TAttribute>(attributes);
+        }
+
+        /// <summary>
+        /// The underlying collection of <typeparamref name="TAttribute"/>s.
+        /// </summary>
+        /// <remarks>Modifying this <see cref="List{TAttribute}"/>s affects what is exposed by this
+        /// <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> instance.</remarks>
```

Comment:
true but suggest instead describing _why_ this property exists. perhaps something like `Intended for use in a non-read-only subclass. Changes to this <see cref="List{TAttribute}"/> will affect all getters that <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> provides.`